src/piadm.sh

```diff
@@ -199,6 +218,104 @@ mount_installmedia() {
+# EFI System Partition (ESP), it can boot iPXE straight off of the
+# ESP.  If it is a BIOS disk, or loader is in the ESP, it will boot to
+# loader like SmartOS, but instead of using "unix" and "boot_archive"
+# from the platform directory, it will instead boot "ipxe.lkrn" and
+# "default.ipxe" from the boot directory.
+#
+# Because of the desire to have a Triton Compute Node possibly boot to
+# a recovery SmartOS, we will never install iPXE on the ESP.
+#
+# "bootfs" has been set already.
+install_ipxe() {
+	cd /${bootfs}
+	if [[ -e platform-ipxe || -e boot-ipxe ]]; then
+		eecho "existing ipxe appears to exist, version: " \
+			$(cat etc/version/ipxe)
+		err "Use 'piadm update ipxe' to upgrade it"
+	fi
+
+	# The "platform" directory for on-disk iPXE is a placeholder.
+	# We put a README (see cat_readme() above) and the string "ipxe"
+	# for the PI-stamp.
+	mkdir -p ./platform-ipxe/etc/version
+	cat_readme > ./platform-ipxe/README
+	echo "ipxe" > ./platform-ipxe/etc/version/platform
+
+	# Entry to this function is predicated by existence of
+	# TRITON_IPXE_LKRN and TRITON_IPXE_ARCHIVE.  Assume
+	# TRITON_IPXE_ETC is there too.
+	mkdir -p etc/version
+	cp -f ${TRITON_IPXE_ETC}/version/ipxe etc/version/.
+	# So 'piadm list' doesn't seize up...
+	echo "ipxe" > etc/version/boot
+	vecho "installing ipxe version: " $(cat etc/version/ipxe)
+
+	mkdir boot-ipxe
+	tar -cf - -C ${TRITON_IPXE_BOOT} . | tar -xf - -C boot-ipxe
+
+	# Populate loader.conf.
+	# XXX KEBE SAYS make loader.conf work for ipxe.
+	# XXX KEBE SAYS FILL ME IN!  AND ONCE FILLED IN CHECK ME!
+	sed 's/headnode="true"/headnode="false"/g' \
+	    < boot-ipxe/loader.conf.tmpl > boot-ipxe/loader.conf
+	echo 'ipxe="true"' >> boot-ipxe/loader.conf
+	echo 'smt_enabled="true"' >> boot-ipxe/loader.conf
+	echo 'console="ttyb,ttya,ttyc,ttyd,text"' >> boot-ipxe/loader.conf
+	echo 'os_console="ttyb"' >> boot-ipxe/loader.conf
+	echo 'fstype="ufs"' >> boot-ipxe/loader.conf
+
+	# Rewhack symlinks
+	ln -s boot-ipxe boot
+	ln -s platform-ipxe platform
+
+	# We're good.
+	return 0
+}
+
+# Remove iPXE on a pool. Called by remove(), which means "bootfs" has
+# been set already and it's PWD.  Unlike SmartOS PIs, we remove iPXE
+# even if it's the next thing to boot.  ./platform-ipxe is guaranteed
+# to be present.
+remove_ipxe() {
+	bootstamp=$1
+
+	if [[ ! -d ./boot-ipxe ]]; then
+		# Catch this now before we remove symlinks
+		corrupt "boot-ipxe is missing for some weird reason"
+	fi
+
+	vecho "Removing ipxe version" $(cat ./etc/version/ipxe)
+
+	/bin/rm -rf ./platform-ipxe ./boot-ipxe ./etc/version/ipxe
+
+	if [[ $bootstamp == "ipxe" ]]; then
+		# Remove the symlinks.
+		# iPXE CAN NOT have mismatched boot and platform symlinks.
+		# It is safe to remove boot as well.
+		rm ./platform ./boot
+		vecho "Removal of active iPXE leaves nothing for $bootfs to find."
+		vecho "Use 'piadm install' to provide something."
+		vecho "If you're using 'piadm update', that will happen now."
```

Comment:
This could also note that external boot media will be required to boot

Reply:
Actually rewhacked this whole section to make "ipxe" more like any other PI and not something super-special (modulo "update").